### PR TITLE
Bug 1858879: Change router's internal endpoint.ID to prevent HAProxy server line collisions

### PR DESCRIPTION
**Pull request by sgreene570**
HAProxy backend stanzas cannot have duplicate `server` lines. Services with multiple port specs that reference the same target port break this rule when a route is created with an integer target port that matches the (duplicate) service target port. Services that meet this criteria should also be counted as 1 (one) endpoint when calculating service endpoint weights and when determining if an HAProxy backend health probe is applicable.

### CODE REVIEW

File: pkg/router/template/router.go [hunk 1]
```diff
@@ -1007,12 +1007,16 @@ func (r *templateRouter) removeRouteInternal(route *routev1.Route) {
...
-	var eps = 0
+	var count = 0
 	svc, ok := r.findMatchingServiceUnit(id)
-	if ok && len(svc.EndpointTable) > eps {
-		eps = len(svc.EndpointTable)
+
+	// Count endpoints that are near identical (same ID)
+	// as the same endpoint to properly calculate endpoint weights
+	endpoints := filterDuplicateEndpointsByID(svc.EndpointTable)
+	if ok && len(endpoints) > count {
```
Comment (Miciah) [inline pkg/router/template/router.go:1016]: If `findMatchingServiceUnit` does not find a match, it returns the zero value, which has empty `EndpointTable`.  This means that checking `ok` is redundant.  Furthermore, after this change, you are using `svc.EndpointTable` before checking `ok`, which looks wrong (if the check _were_ needed—for example, if `svc` were a pointer value—then `svc.EndpointTable` would be a nil-pointer dereference).  How about doing away with `ok` entirely?

File: pkg/router/template/template_helper.go [hunk 2]
```diff
@@ -149,6 +153,20 @@ func endpointsForAlias(alias ServiceAliasConfig, svc ServiceUnit) []Endpoint {
...
 }
 
+// filterDuplicateEndpointsByID returns an array of Endpoints
+// with unique IDs
+func filterDuplicateEndpointsByID(eps []Endpoint) []Endpoint {
+	contains := make(map[string]bool)
+	unique := []Endpoint{}
+	for _, ep := range eps {
+		if _, value := contains[ep.ID]; !value {
+			contains[ep.ID] = true
```
[the remark this answers is not in the document]
Comment (sgreene570) [reply pkg/router/template/template_helper.go:163]: TIL!

**Comment by sgreene570** [comment]
> Would you mind adding unit tests for `calculateServiceWeights` and for `processEndpointsForAlias`?

Sure, I will add unit tests for those 2 functions. Thanks for the suggestion!

> What if a route specifies two services that have an endpoint in common? (Maybe that is outside the scope of this fix.)

Are you concerned about the service weights or HAProxy breaking in this situation? A route that specifies two services with a common endpoint will not break HAProxy with or without this change, since each service is given a separate backend stanza.

This PR does not change how endpoints common across multiple services will be weighted, since duplicate endpoints are filtered on a per-service basis. Are you suggesting that we change how services with common endpoints weight each of their endpoints? If so, I think that is out of scope of this particular bugfix.

**Comment by Miciah** [comment]
> Are you concerned about the endpoint weights being incorrect or HAProxy breaking in this situation?

I meant the latter (HAProxy's breaking if a route has two services with a common endpoint, same as it breaks if a route has one service with two identical endpoints).  

> A route that specifies two services with a common endpoint will not break HAProxy with or without this change, since each service is given a separate backend stanza.

Are you sure that each service is given a separate `backend` stanza?  A route with two services is translated to a service unit alias with two associated service units:  https://github.com/openshift/router/blob/4a6284e05280e1e565d4d2dccd52ed371c6f4127/pkg/router/template/router.go#L854-L856 https://github.com/openshift/router/blob/4a6284e05280e1e565d4d2dccd52ed371c6f4127/pkg/router/template/router.go#L1171-L1177
The template produces a `backend` stanza for each service unit alias:  https://github.com/openshift/router/blob/4a6284e05280e1e565d4d2dccd52ed371c6f4127/images/router/haproxy/conf/haproxy-config.template#L435-L440
And under that `backend` stanza, the template produces a `server` stanza for each endpoint of each service unit associated with the service unit alias:  https://github.com/openshift/router/blob/4a6284e05280e1e565d4d2dccd52ed371c6f4127/images/router/haproxy/conf/haproxy-config.template#L559-L563
Line 559 is iterating over the service unit, and line 562 is iterating over the endpoints of the service unit.  So it looks like two services with the same endpoint would produce two `server` stanzas, right?

That said, in tracing through this code, I realized that `$endpoint.ID` includes the name of the Kubernetes endpoints object, which means that the endpoint id will be different for two endpoints that come from different services, even if the endpoints are the same:  https://github.com/openshift/router/blob/4a6284e05280e1e565d4d2dccd52ed371c6f4127/pkg/router/template/plugin.go#L325

So this last discovery means that having a route with two services that have a common endpoint should not break anything, as the server name (= endpoint id) will differ.

However, this last discovery also raises a question:  Would it be an alternative (and simpler) solution to change the endpoint id to include the port name (`p.Name`)?

**Comment by sgreene570** [comment]
> A route with two services is translated to a service unit alias with two associated service units

Ah, there's the missing piece that tripped me up. Thanks for the detailed explanation! I will ear mark this PR for future reference :smile: 

> Would it be an alternative (and simpler) solution to change the endpoint id to include the port name (p.Name)?

Yes, absolutely. I prefer this approach more, since endpoints would not need to be filtered out at all, and weight calculation code does not need to be modified. Also, `p.Name` collisions are not possible, from my understanding. Thank you for this suggestion!

I'll go ahead and make the change to the ep.ID string format, along with add in the unit tests mentioned earlier (not as important with this new approach, but still should be covered from this point forward, so I'll take care of that!).

**Comment by sgreene570** [comment]
```
Unable to connect to the server: dial tcp: lookup api.ci-op-h9sw744l-8f0fe.origin-ci-int-gce.dev.openshift.com on 10.142.0.107:53: no such host
```
/retest

File: pkg/router/template/template_helper_test.go [hunk 3]
```diff
@@ -701,3 +702,64 @@ func TestGetPrimaryAliasKey(t *testing.T) {
...
+			name:           "3 basic endpoints with different Port string",
+			preferPort:     "foo",
+			endpoints:      []Endpoint{ep1, ep2, ep3},
+			expectedLength: 2,
+		},
+	}
+
+	for _, tc := range testCases {
+		alias.PreferPort = tc.preferPort
+		router.AddEndpoints(suKey, tc.endpoints)
```
[the remark this answers is not in the document]
Comment (sgreene570) [reply pkg/router/template/template_helper_test.go:752]: Ah, yea, definitely makes sense to copy the endpoints table first. Thanks for catching this!!

[the remark this answers is not in the document]
Comment (sgreene570) [reply pkg/router/template/template_helper_test.go]: Fixed.

**Comment by sgreene570** [comment]
e2e tests failing on metrics failures from endpoint ID changes. Will fix.
/hold

**Comment by Miciah** [comment]
Last couple of e2e runs failed on [BZ#1873728](https://bugzilla.redhat.com/show_bug.cgi?id=1873728).

**Comment by Miciah** [comment]
/hold
while we investigate [BZ#1874278](https://bugzilla.redhat.com/show_bug.cgi?id=1874278).

**Comment by sgreene570** [comment]
`error: failed to push image `
/retest

**Pull request merged**